File: upstream/ray-gauges/packages/cp-lp-escrow-sdk/src/personalPosition.ts
import { BN, web3 } from "@coral-xyz/anchor"
import { accounts, instructions, PROGRAM_ID } from "@raygauge/cp-lp-escrow-gen"
import { CpLpEscrowPDA } from "@raygauge/cp-lp-escrow-pda"
import { PreciseNumber } from "@raygauge/number"
import { TOKEN_PROGRAM_ID } from "@solana/spl-token"

export interface CpLpEscrowPersonalPositionLoadArgs {
  connection: web3.Connection
  poolId: web3.PublicKey
  owner: web3.PublicKey
  programId?: web3.PublicKey
}

type CpLpEscrowPersonalPositionConfig = {
  programId?: web3.PublicKey
  poolId: web3.PublicKey
  selfAddress: web3.PublicKey
  state: {
    account: accounts.PersonalPosition
  }
}

/** SDK for interacting with a personal position */
export class CpLpEscrowPersonalPosition {
  programId: web3.PublicKey
  pda: CpLpEscrowPDA

  constructor(private config: CpLpEscrowPersonalPositionConfig) {
    this.programId = config.programId || PROGRAM_ID
    this.pda = new CpLpEscrowPDA(this.programId)
  }

  static async load(args: CpLpEscrowPersonalPositionLoadArgs) {
    const pda = new CpLpEscrowPDA(args.programId)
    const address = pda.personalPosition({ poolId: args.poolId, owner: args.owner })
    const state = await loadState(args.connection, address)
    return new CpLpEscrowPersonalPosition({
      programId: args.programId,
      poolId: args.poolId,
      selfAddress: address,
      state,
    })
  }

  async reload(connection: web3.Connection) {
    const state = await loadState(connection, this.config.selfAddress)
    this.config.state = state
  }

  /** Deposit LP tokens into the escrow */
  depositIx({ amount, lpSrc }: { amount: bigint; lpSrc: web3.PublicKey }) {
    return instructions.deposit(
      { amount: new BN(amount.toString()) },
      {
        owner: this.owner,
        timeTracker: this.timeTracker,
        personalPosition: this.selfAddress,
        escrow: this.escrowAddress,
        lpSrc,
        tokenProgram: TOKEN_PROGRAM_ID,
      },
    )
  }

  /** Withdraw LP tokens from the escrow */
  withdrawIx({ amount, lpDst }: { amount: bigint; lpDst: web3.PublicKey }) {
    return instructions.withdraw(
      { amount: new BN(amount.toString()) },
      {
        owner: this.owner,
        timeTracker: this.timeTracker,
        personalPosition: this.selfAddress,
        escrow: this.escrowAddress,
        lpDst,
        tokenProgram: TOKEN_PROGRAM_ID,
      },
    )
  }

  /** Sync the personal position's earned time units */
  syncIx() {
    return instructions.updatePersonalPosition({
      personalPosition: this.selfAddress,
      timeTracker: this.timeTracker,
    })
  }

  get state() {
    return this.config.state
  }

  /** Address of the escrow account for LP tokens */
  get escrowAddress() {
    return this.pda.lpEscrow({ poolId: this.poolId })
  }

  get owner() {
    return this.state.account.owner
  }

  /** Address of the pool */
  get poolId() {
    return this.config.poolId
  }

  /** Address of the time tracker */
  get timeTracker() {
    return this.state.account.timeTracker
  }

  /** Laste seen reward index */
  get lastSeenIndex(): string {
    return PreciseNumber.fromRaw(this.state.account.lastSeenIndex.val).valueString
  }

  /** High-precision number representation of earned units */
  get earnedTimeUnits(): string {
    return PreciseNumber.fromRaw(this.state.account.earnedTimeUnits.val).valueString
  }

  /** Balance of LP tokens */
  get balance(): bigint {
    return BigInt(this.state.account.amount.toString())
  }

  /** Address of the personal position */
  get selfAddress(): web3.PublicKey {
    return this.config.selfAddress
  }

  toJson(): CpLpEscrowPersonalPositionJson {
    return toJson(this)
  }
}

async function loadState(connection: web3.Connection, address: web3.PublicKey) {
  const account = await accounts.PersonalPosition.fetch(connection, address)
  return { account }
}

function toJson(sdk: CpLpEscrowPersonalPosition): CpLpEscrowPersonalPositionJson {
  return {
    selfAddress: sdk.selfAddress.toBase58(),
    owner: sdk.owner.toBase58(),
    balance: sdk.balance.toString(),
    lastSeenIndex: sdk.lastSeenIndex,
    earnedTimeUnits: sdk.earnedTimeUnits,
    timeTracker: sdk.timeTracker.toBase58(),
    poolId: sdk.poolId.toBase58(),
  }
}

export interface CpLpEscrowPersonalPositionJson {
  /** Address of the personal position */
  selfAddress: string
  /** Address of the owner */
  owner: string
  /** Balance of LP tokens */
  balance: string
  /** Laste seen reward index */
  lastSeenIndex: string
  /** High-precision number representation of earned units */
  earnedTimeUnits: string
  /** Address of the time tracker */
  timeTracker: string
  /** Address of the pool */
  poolId: string
}

File: upstream/ray-gauges/sol/programs/cp-lp-escrow/src/state/personal_position.rs
use anchor_lang::prelude::*;
use precise_number::Number;

use crate::state::NumberRaw;

#[account]
pub struct PersonalPosition {
    pub owner: Pubkey,

    /// Amount of liquidity
    amount: u64,

    /// Link to time tracker
    pub time_tracker: Pubkey,

    pub last_seen_index: NumberRaw,

    /// High-precision number representation of earned units
    pub earned_time_units: NumberRaw,
}

impl PersonalPosition {
    pub const SIZE: usize =
        // account discriminator
        8 +
        // owner
        32 +
        // amount
        8 +
        // time_tracker
        32 +
        // last_seen_index
        NumberRaw::SIZE +
        // earned_time_units
        NumberRaw::SIZE;

    /// Increase amount of LP tokens in the personal position
    pub fn inc_amount(&mut self, cur_index: Number, amount: u64) {
        // update the earned time units
        self.update(cur_index);

        self.amount = self
            .amount
            .checked_add(amount)
            .expect("personal position overflow");
    }

    /// Decrease amount of LP tokens in the personal position
    pub fn dec_amount(&mut self, cur_index: Number, amount: u64) {
        // update the earned time units
        self.update(cur_index);

        self.amount = self
            .amount
            .checked_sub(amount)
            .expect("personal position underflow");
    }

    /// Update the earned time units
    pub fn update(&mut self, cur_index: Number) {
        assert!(
            cur_index >= self.last_seen_index.into(),
            "index must be non-decreasing"
        );

        let delta_i = cur_index - self.last_seen_index.into();

        if delta_i == Number::ZERO {
            return;
        }

        let earned = Number::from_natural_u64(self.amount) * delta_i;

        let cur_earned_time_units: Number = self.earned_time_units.into();
        let new_earned_time_units = cur_earned_time_units + earned;

        self.earned_time_units = new_earned_time_units.into();
        self.last_seen_index = cur_index.into();
    }
}
